### PR TITLE
Interactive rescoring and reclustering

### LINKED ISSUE #631: Export plotly data and layout as json
#611 adds a nice webpage with plotly graphs and tables. To embed these in another page (https://github.com/i-VRESSE/bartended-haddock3) we could use https://plotly.com/javascript/react/. For that, it would be helpful to not just output the reports as html, but also as json.

Perhaps the nicest option would be that the report.html also uses report.json to get the figure layout and data.

It might be necessary to have a separate json file for each plotly object.

### LINKED ISSUE #686: interactive re-scoring
To make the rescoring interactive on the haddock3-webapp we need a quick way to rescore a caprieval step and to re-analyse the data.
The first approach will be a quick CLI that only modifies CAPRI data, putting them in a `X_caprieval_interactive` folder, followed by the standard haddock3-analyse CLI for the analysis, which will need a minor modification to handle these new folders.
This approach must be properly benchmarked in terms of execution speed.
For more info: i-VRESSE/haddock3-webapp/issues/50
Discussed with @sverhoeven

### LINKED ISSUE #739: clarify naming in clustrmsd and clustfcc
naming in clustering modules should follow #685, namely
```
tolerance -> clust_cutoff (if numerical value)
tolerance -> n_clusters (if we want a certain number of clusters)
threshold -> min_population
```

### LINKED ISSUE #786: Reusable plot data
**Is your feature request related to a problem? Please describe.**
Haddock3 can generate interactive plots in html files with plotly.
In the haddock3 webapp we want to show those plots in different ways.
It would be helpful if the plotly data could be retrieved in an easy way.

**Describe the solution you'd like**

Would like to see a html file from which it is easy to parse the plotly data

```html
<div>
    <script type="text/javascript">window.PlotlyConfig = { MathJaxConfig: 'local' };</script>
    <script src="https://cdn.plot.ly/plotly-2.16.1.min.js"></script>
    <div id="6ce37419-297c-401e-860d-61704c0795b8" class="plotly-graph-div" style="height:800px; width:1000px;">
    </div>
    <script id="plotly1" type="application/json">
        {
            "data": [
              {
                "x": [
                  1999,
                  2000,
                  2001,
                  2002
                ],
                "y": [
                  10,
                  15,
                  13,
                  17
                ],
                "type": "scatter"
              }
            ],
            "layout": {
              "title": "Sales Growth",
              "xaxis": {
                "title": "Year",
                "showgrid": false,
                "zeroline": false
              },
              "yaxis": {
                "title": "Percent",
                "showline": false
              }
            }
          }
    </script>
    <script type="text/javascript">
        const { data, layout } = JSON.parse(document.getElementById("plotly1").text)
        window.PLOTLYENV = window.PLOTLYENV || {};
        if (document.getElementById("6ce37419-297c-401e-860d-61704c0795b8")) {
            Plotly.newPlot(
                "6ce37419-297c-401e-860d-61704c0795b8",
                data,
                layout,
                { responsive: true },
            );
        }
    </script>
</div>
```

The JSON can be generated using the `fig.to_json()` method.

In the webapp I can then parse the HTML to get the text of the script tag with id `plotly1`.

**Describe alternatives you've considered**
- Now the webapp parses the tsv files to generates the plotly by duplicating code from Python to JS.
- Having separate JSON file, but then you need to run a web server to workaround cors

**Additional context**

### PR DESCRIPTION
**Pull request by mgiulini**
You are about to submit a new Pull Request. Before continuing make sure you read the [contributing guidelines](CONTRIBUTING.md) and that you comply with the following criteria:

- [ ] You have sticked to Python. Please talk to us before adding other programming languages to HADDOCK3
- [ ] Your PR is about CNS
- [ ] Your code is well documented: proper docstrings and explanatory comments for those tricky parts
- [ ] You structured the code into small functions as much as possible. You can use classes if there is a (state) purpose
- [ ] Your code follows our coding style
- [ ] You wrote tests for the new code
- [ ] `tox` tests pass. *Run `tox` command inside the repository folder*
- [ ] `-test.cfg` examples execute without errors. *Inside `examples/` run `python run_tests.py -b`*
- [ ] PR does not add any dependencies, unless permission granted by the HADDOCK team
- [ ] PR does not break licensing
- [ ] Your PR is about writing documentation for already existing code :fire:
- [ ] Your PR is about writing tests for already existing code :godmode:

---

Closes #739 
Closes #686 
Closes #786 
Closes #631

Briefly, this PR solves the following:

- clarifies the naming in the clustering modules (and related examples)
- breaks down the clustering modules in small functions
- creates a new CLI (haddock3-re) dedicated to interactive tasks. This command can be launched with three subcommands (so far): `score`, `clustfcc`, and `clustrmsd`
- embeds json schemas into the output html files

As an example, the rescoring CLI can be applied to a caprieval step as follows:
`haddock3-re score run1-ranairCDR-test/02_caprieval/ -e 0.3`
this will change the weight of the electrostatic component to 0.3, keeping the other weights to the ones that were originally used by haddock3. A new directory will be created in the workflow, named `02_caprieval_interactive` in this case, that will be used by the haddock3 webapp to allow for interactive rescoring in the browser.

The same logic applies to the two reclustering subcommands. They have the feature that they can trace back the closest caprieval folder and update it, thus allowing to rerun the analysis with the modified clusters. If a CNS module is found in between, the search is interrupted.

As an example, take a workflow with
0_topoaa
1_rigidbody
2_caprieval
3_clustfcc

running `haddock3-re clustfcc workflow/3_clustfcc ` will create a folder named `3_clustfcc_interactive`, that contains the data present in 2_caprieval with the new clustering information. if there was a flexref folder between 2_caprieval and 3_clustfcc the caprieval data would not be touched.

@sverhoeven cannot add you as a reviewer, but please have a look at this PR.

### CODE REVIEW

**Review by amjjbonvin — COMMENTED** [review]
We should check the haddock3 tutorials in case we need to change the parameter naming there as well.

**Comment by mgiulini** [comment]
> We should check the haddock3 tutorials in case we need to change the parameter naming there as well.

good point, I added a related issue on the website repo.

File: examples/docking-protein-glycan/docking-protein-glycan-full.cfg [hunk 1]
```diff
@@ -70,7 +70,7 @@ resdic_B = [ 1 , 2 , 3 , 4 ]
 criterion = 'distance'
 linkage = 'average'
 # full example, 4 models should be present in a cluster
-threshold = 4
+min_population = 4
 tolerance = 2.5
```
Comment (VGPReys) [inline examples/docking-protein-glycan/docking-protein-glycan-full.cfg:74]: shouldn't it be `clust_cutoff = 2.5` ?

Comment (mgiulini) [reply examples/docking-protein-glycan/docking-protein-glycan-full.cfg:74]: yes it should, good catch!

**Comment by VGPReys** [comment]
I did a PR #778 to be merged to this PR

**Comment by sverhoeven** [comment]
Not all examples have their clustr params renamed completely. For example https://github.com/haddocking/haddock3/blob/337b96a1900c907297bd5c18983bf8a5e8b215e3/examples/docking-protein-glycan/docking-protein-glycan-full.cfg#L50

Is there a test or command to check that the recipes in examples directory are valid?

**Comment by mgiulini** [comment]
> Not all examples have their clustr params renamed completely. For example
> 
> https://github.com/haddocking/haddock3/blob/337b96a1900c907297bd5c18983bf8a5e8b215e3/examples/docking-protein-glycan/docking-protein-glycan-full.cfg#L50
> 
> Is there a test or command to check that the recipes in examples directory are valid?

good catch, thanks! they should be OK now. They were mainly coming from a merging of another branch

we don't have that validation yet, but it would be a very nice integration test..let me think about it!

**Comment by VGPReys** [comment]
> we don't have that validation yet, but it would be a very nice integration test..let me think about it!

Great idea !
The functions validating the config files should do the trick.
Should look like the function `setup_run(cfg_filepath)`, present in `gear/prepare_run.py`

**Comment by sverhoeven** [comment]
re-re-running clustrmsd, using docking-protein-glycan-test.cfg example, also goes out of sync

```
wc -l 10_clustrmsd_interactive/capri_clt.tsv analysis/10_clustrmsd_interactive_analysis/capri_clt.tsv
   4 10_clustrmsd_interactive/capri_clt.tsv
  10 analysis/10_clustrmsd_interactive_analysis/capri_clt.tsv
```

**Pull request merged**